File: 279.Perfect-Squares.py
# https://leetcode.com/problems/perfect-squares/
#
# algorithms
# Medium (40.82%)
# Total Accepted:    162,998
# Total Submissions: 399,357


class Solution(object):

    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        if self.is_perfect_num(n):
            return 1

        perfect_num_arr = []

        i = 1
        while i * i < n:
            perfect_num_arr += (i ** 2),
            i += 1

        dp = [float('inf')] * (n + 1)
        for p_n in perfect_num_arr:
            dp[p_n] = 1

        for i in xrange(2, n + 1):
            cur_min = float('inf')
            for p_n in perfect_num_arr:
                if i - p_n <= 0:
                    break
                cur_min = min(cur_min, dp[i - p_n])

            dp[i] = min(dp[i], cur_min + 1)

        return dp[-1]

    def is_perfect_num(self, n):
        return int(math.sqrt(n)) ** 2 == n

File: 279.Perfect-Squares.go
// https://leetcode.com/problems/perfect-squares/
//
// algorithms
// Medium (41.65%)
// Total Accepted:    176,708
// Total Submissions: 424,280

package leetcode

func numSquares(n int) int {
	var squares []int
	idx := 1

	for {
		tmp := idx * idx
		if tmp == n {
			return 1
		}
		if tmp > n {
			break
		}
		squares = append([]int{tmp}, squares...)
		idx++
	}

	dp := make([]int, n+1)
	for _, s := range squares {
		dp[s] = 1
	}

	for i := 1; i < n; i++ {
		for _, s := range squares {
			if i+s <= n {
				if dp[i+s] == 0 {
					dp[i+s] = 1 + dp[i]
				} else {
					dp[i+s] = min(dp[i+s], 1+dp[i])
				}
			}
		}
	}

	return dp[n]
}

func min(a, b int) int {
	if a < b {
		return a
	}

	return b
}
